jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -591,46 +647,151 @@ public String toString()
         FAILURE
     }
 
-    private class Decoder extends IteratingNestedCallback
+    private class ContentListeners
+    {
+        private final Map<Object, Long> demands = new ConcurrentHashMap<>();
+        private final LongConsumer demand = HttpReceiver.this::demand;
+        private final List<Response.DemandedContentListener> listeners;
+
+        private ContentListeners(List<Response.ResponseListener> responseListeners)
+        {
+            listeners = responseListeners.stream()
+                .filter(Response.DemandedContentListener.class::isInstance)
+                .map(Response.DemandedContentListener.class::cast)
+                .collect(Collectors.toList());
+        }
+
+        private boolean isEmpty()
+        {
+            return listeners.isEmpty();
+        }
+
+        private void notifyBeforeContent(HttpResponse response)
+        {
+            if (isEmpty())
+            {
+                // If no listeners, we want to proceed and consume any content.
+                demand.accept(1);
+            }
+            else
+            {
+                ResponseNotifier notifier = getHttpDestination().getResponseNotifier();
+                notifier.notifyBeforeContent(response, this::demand, listeners);
+            }
+        }
+
+        private void notifyContent(HttpResponse response, ByteBuffer buffer, Callback callback)
+        {
+            if (hasManyListeners())
+                demands.replaceAll((k, v) -> v - 1);
+            HttpReceiver.this.demand(d -> d - 1);
+            ResponseNotifier notifier = getHttpDestination().getResponseNotifier();
+            notifier.notifyContent(response, this::demand, buffer, callback, listeners);
+        }
+
+        private void demand(Object context, long value)
+        {
+            if (hasManyListeners())
+                accept(context, value);
+            else
+                demand.accept(value);
+        }
+
+        private boolean hasManyListeners()
+        {
+            return listeners.size() > 1;
+        }
+
+        private void accept(Object context, long value)
+        {
+            demands.merge(context, value, MathUtils::cappedAdd);
+            if (demands.size() == listeners.size())
+            {
+                long minDemand = Long.MAX_VALUE;
+                for (Long demand : demands.values())
+                {
+                    if (demand < minDemand)
+                        minDemand = demand;
+                }
+                if (minDemand > 0)
```

Comment:
To me it feels like we should now subtract minDemand from all entries in the map, removing zero results from the map, so that the size check at line 708 above is again meaningful.      If you don't remove 0 entries, then all future calls to accept will always have `(demands.size() == listeners.size())` and you will always iterate over the map, even when you could know you didn't have enough.

Reply:
Semantic-wise, to me `accept()` is where I _increment_ the demand, so feels weird to decrement it here. The demand is decremented just before notifying.

State-wise, I like the fact that the `demands` map holds the current state of the demand for each listener.
In your case, I would have to reconcile the global demand with the content of the map; missing listeners in the map may have demand if there is global demand and existing listener have their demand increased by the global demand.

With 3 listeners and 5 chunks, and the "normal" pattern of calling `demand(1)`, in my case the code does 3*5=15 `put/replace` and then 1+3*4=13 iterations; in your case I do 3 `put` + 2 iterations (1 to calculate the minimum + 1 to remove the listeners) + 3 `remove` per chunk so it's 15 `put()` + 10 iterations + 15 `remove()`.
With 50 chunks, excluding the `put/replace` operation that is the same for the 2 solutions, it's 1+3*49=148 iterations versus 2*50=100 iterations + 3*50=150 `remove`.

I doubt there will be more than a handful of listeners, so I would keep my version unless I got the above wrong.

Reply:
I still really don't like this.  The `demands.size() == listeners.size()` test is just wrong because it's only meaningful for the first loop. 

Let's walk through your example with this code:
 - listener A does demand(1), {A=1}, size test fails
 - listener B does demand(1), {A=1,B=1}, size test fails
 - listener C does demand(1), {A=1,B=1,C=1}, size test passes:
    + iterate to find minimum
    + send real demand 1
 - notifyContent is called
    + iterates setting -1 on ABC demands so we have {A=0,B=0,C=0}
    + listeners are notified (the below may happen within that call scope)
 - listener A does demand(1), size test passes on  {A=1,B=0,C=0}
    + iterate to find minimum of 0  - WASTED EFFORT!
 - listener B does demand(1), size test passes on  {A=1,B=1,C=0}
    + iterate to find minimum of 0 - WASTED EFFORT!
 - listener C does demand(1), {A=1,B=1,C=1}, size test passes:
    + iterate to find minimum
    + send real demand 1

So there is wasted effort and the second chunk is handled differently from the first chunk.   Plus the worst case is nasty because iterations growth is non linear: Iterations = 1 + chunks + (chunks-1) * listeners


With my proposal:
 - listener A does demand(1), {A=1}, size test fails
 - listener B does demand(1), {A=1,B=1}, size test fails
 - listener C does demand(1), {A=1,B=1,C=1}, size test passes:
    + iterate to find minimum of 1
    + iterate to do 1-1==0 remove resulting in {}
    + send real demand 1
 - notifyContent is called
    + listeners are notified (the below may happen within that call scope)
 - listener A does demand(1), {A=1}, size test fails
 - listener B does demand(1), {A=1,B=1}, size test fails
 - listener C does demand(1), {A=1,B=1,C=1}, size test passes:
    + iterate to find minimum of 1
    + iterate to do 1-1==0 remove resulting in {}
    + send real demand 1

Each chunk is processed exactly the same. Iterations grown is linear: Iterations = 2 * chunks